Properly construct instance of LateBoundBatchSpanProcessor

geoand opened the pull request:
This is now done entirely in a CDI producer method which is then fully initialized when OpenTelemetry autoconfigures itself

Fixes: #29987 (which with newer Quarkus versions is even [more problematic](https://github.com/quarkusio/quarkus/issues/29987#issuecomment-1621594578))

Detailed explanation:

* The first commit fixes the case where the DataSource cannot properly be constructed because the `OpenTelemetry` has not yet been created by CDI. This could happen because there was not declaration of the need for that bean when constructing the DataSource bean, only programmatic lookup.
* The second commits actually fixes the issue with the missing delegate in a pretty straightforward way: We move the creation of `LateBoundBatchSpanProcessor` to a synthetic bean - prior to this commit the `LateBoundBatchSpanProcessor` bean was created from a CDI producer but was then "updated" by a recorder. These two operations are obviously not "atomic", meaning that the `OpenTelemetry` bean that consumed `LateBoundBatchSpanProcessor` would see the bean in the intermediate state - where it was created but not updated.
    * This commit also includes the removal of various intermediate build items whose only purpose was to pass data between the various steps. These are no longer necessary, since everything can be accomplished with the combination of CDI and the `BeanContainer` in recorder methds.
    * Furthermore, this commit makes sure that we no longer use `Arc.container()...` in recorder methods as these are extremely sensitive to timing issues - by utilizing synthetic beans and the `BeanContainer` in recorder methods we overcome this problem.
* The third commit builds on the second one to make the configuration of the `SpanProcessor`s of `OpenTelemetry` even more reasonable. We essentially only include  `LateBoundBatchSpanProcessor` in the `OpenTelemetry` bean if a delegate has been set. 

_Future work_: The creation of the `OpenTelemetry` can be done via a synthetic bean as this allows us complete control over the injection points and does not suffer from any bean creating timing issues (assuming of course the creation is done properly).
This would essentially take what is not done in `OpenTelemetryProducer` and move it into a recorder method that includes all the creation logic (including the logic that currently resides in the synthetic bean for `LateBoundBatchSpanProcessor`)

geoand commented:
Correct, the bean initialization is not forced at build time anymore

brunobat commented:
This makes everything much simpler. Not sure about the increased startup time... Most likely not noticeable.
I need to check possible implications on the OpenTelemetryProducer, but I won't have time today.

geoand commented:
No rush!

gsmet commented:
If it gets merged, should it be backported?

geoand commented:
I dunno, I'm on the fence...

radcortez commented:
I think we can hold this until 3.3.

brunobat commented on extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/tracing/TracerRecorder.java:
```diff
@@ -27,24 +17,13 @@ public class TracerRecorder {
-    public RuntimeValue<Optional<IdGenerator>> createIdGenerator() {
-        BeanManager beanManager = Arc.container().beanManager();
-
-        Instance<IdGenerator> idGenerator = beanManager.createInstance()
```
This might break idGeneration. We don't have a proper test. Will come up with something, to make sure.

geoand replied:
Wouldn't we already have this problem in the current version?

brunobat reviewed (changes requested):
I need to create a new test before approving this PR.

brunobat commented:
@radcortez @gsmet I'd like this back-ported to 3.2.* because these changes affect core parts of the extension and 3.2.* will run for a long time, potentially causing complex merges in the future.

geoand commented:
If we want this backported, we probably should do it as soon as possible instead of waiting for late 3.2.z releases

brunobat reviewed (approved):
Let's be pragmatic then. 
I'll not be able to do this while at the RivieraDev conf. 
Any iD generator issue can be dealt in this new issue: https://github.com/quarkusio/quarkus/issues/34668

geoand commented:
I'm going to have to redo this one unfortunately since there are plenty of conflicts... 

Furthermore, this will conflict with #34647

geoand commented:
Here is my proposal for this one:

I redo it after https://github.com/quarkusio/quarkus/pull/34647 is in, in order to avoid additional conflicts.
For 3.2, I open a new PR specifically targetting the 3.2 branch. 

WDYT?

radcortez commented:
Sure.

geoand committed:
Properly construct instance of LateBoundBatchSpanProcessor

This is now done entirely in a CDI producer method which is then fully initialized when OpenTelemetry autoconfigures itself

Fixes: #29987 (which with newer Quarkus versions is even more problematic)

geoand commented:
Should be working now

geoand commented:
But there is another thing I want to improve if possible

geoand committed:
Further improve the construction of LateBoundBatchSpanProcessor

geoand commented:
Should be good to go now

geoand commented:
In future, I would like to move the creation of the `OpenTelemetry` bean out of `OpenTelemetryProducer` and into a synthetic bean, but I'll leave that for some other time because it's a lot of mechanical work that is not absolutely necessary at the moment

geoand commented:
Oh, I didn't know about those tests...

geoand commented:
The JDBC telemetry stuff is pretty broken (it only worked by accident before) so I need to see what can be done...

brunobat commented:
Those tests are a bit brittle but beware that they end up finding a lot of problems.

geoand commented:
The tests themselves are great and they indeed show a real problem here.

geoand committed:
Ensure that DataSource beans dependencies are initialized before the datasources

brunobat commented:
Ok @geoand. The interdependencies with vert.x in `InstrumentationRecorder.createTraces(..)`
Will an OTel Producer as a synthetic bean work with dev mode reload?

geoand commented:
All tests pass so this is good from my end.

brunobat reviewed (approved):
Thanks @geoand !
